index.js

```diff
@@ -637,10 +666,47 @@ SendStream.prototype.sendFile = function sendFile(path) {
     fs.stat(p, function (err, stat) {
       if (err) return next(err)
       if (stat.isDirectory()) return next()
-      self.emit('file', p, stat)
-      self.send(p, stat)
+      checkPrecompressionAndSendFile(p, stat)
     })
   }
+
+  function checkPrecompressionAndSendFile(p, stat) {
+    self.emit('file', p, stat)
+    if (!self._precompressionFormats) return self.send(p, stat)
+
+    var state = {
+      contents: [],
+      extensionsToCheck: self._precompressionFormats.length
+    }
+
+    self._precompressionFormats.forEach(function (format) {
+      debug('stat "%s%s"', p, format.extension);
+      fs.stat(p + format.extension, function onstat(err, contentStat) {
+        if (!err) state.contents.push({ext: format.extension, encoding: format.encoding, contentStat: contentStat})
+        if (--state.extensionsToCheck == 0) sendPreferredContent(p, stat, state.contents)
+      })
+    })
+  }
+
+  function sendPreferredContent(p, stat, contents) {
+    if (contents.length) self.res.setHeader('Vary', 'Accept-Encoding')
```

Comment:
This will overwrite any other Vary already set. Let's append this to the field name if already set instead of overwriting.